gcc/rust/checks/errors/rust-hir-pattern-analysis.cc

```diff
@@ -0,0 +1,1554 @@
+  if (matrix.get_place_infos ().empty ())
+    return WitnessMatrix::make_empty ();
+
+  std::vector<Constructor> heads;
+  for (auto head : matrix.heads ())
+    heads.push_back (head.ctor ());
+
+  // TODO: not sure missing ctors need to be calculated
+  auto ctors_and_missings
+    = split_constructors (heads, matrix.get_place_infos ().at (0));
+  std::set<Constructor> ctors = ctors_and_missings.first;
+  std::set<Constructor> missings = ctors_and_missings.second;
+
+  WitnessMatrix ret = WitnessMatrix::make_empty ();
+  for (auto &ctor : ctors)
+    {
+      rust_debug ("specialize with %s", ctor.to_string ().c_str ());
+      // TODO: Instead of creating new matrix, we can change the original matrix
+      // and use it for sub-pattern matching. It will significantly reduce
+      // memory usage.
+      Matrix spec_matrix = matrix.specialize (ctor);
+
+      WitnessMatrix witness
+	= compute_exhaustiveness_and_usefulness (ctx, spec_matrix);
+
+      TyTy::BaseType *ty = matrix.get_place_infos ().at (0).get_type ();
+      witness.apply_constructor (ctor, missings, ty);
+      ret.extend (witness);
+    }
+
+  return ret;
+}
+
+static void
+emit_exhaustiveness_error (Resolver::TypeCheckContext *ctx,
+			   HIR::MatchExpr &expr, WitnessMatrix &witness)
+{
+  TyTy::BaseType *scrutinee_ty;
+  bool ok = ctx->lookup_type (
+    expr.get_scrutinee_expr ()->get_mappings ().get_hirid (), &scrutinee_ty);
+  rust_assert (ok);
+
+  if (!witness.empty ())
+    {
+      std::stringstream buf;
+      for (size_t i = 0; i < witness.get_stacks ().size (); i++)
+	{
+	  auto &stack = witness.get_stacks ().at (i);
+	  WitnessPat w = WitnessPat::make_wildcard (scrutinee_ty);
+	  if (!stack.empty ())
+	    w = stack.at (0);
+
+	  rust_debug ("Witness[%d]: %s", (int) i, w.to_string ().c_str ());
+	  buf << "'" << w.to_string () << "'";
+	  if (i != witness.get_stacks ().size () - 1)
+	    buf << " and ";
+	}
+      rust_error_at (expr.get_scrutinee_expr ()->get_locus (),
+		     "non-exhaustive patterns: %s not covered",
+		     buf.str ().c_str ());
+    }
+  else
+    {
+      rust_debug ("no witness found");
+    }
+}
+
+// Entry point for computing match usefulness and check exhaustiveness
+void
+check_match_usefulness (Resolver::TypeCheckContext *ctx,
+			TyTy::BaseType *scrutinee_ty, HIR::MatchExpr &expr)
+{
+  // Lower the arms to a more convenient representation.
+  std::vector<MatrixRow> rows;
+  for (auto &arm : expr.get_match_cases ())
+    {
+      PatStack pats;
+      MatchArm lowered = lower_arm (ctx, arm, scrutinee_ty);
+      PatOrWild pat = PatOrWild::make_pattern (lowered.get_pat ());
+      pats.push (pat);
+      rows.push_back (MatrixRow (pats, lowered.has_guard ()));
+    }
+
+  std::vector<PlaceInfo> place_infos = {{PlaceInfo (scrutinee_ty)}};
+  Matrix matrix{rows, place_infos};
+
+  WitnessMatrix witness = compute_exhaustiveness_and_usefulness (ctx, matrix);
+
+  emit_exhaustiveness_error (ctx, expr, witness);
+}
```

Comment:
This is an entry point function that is called for each match expression.
It converts match all match arms to a [matrix](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_pattern_analysis/usefulness/index.html#computing-usefulness-and-exhaustiveness-in-one-go) which we compute witness (counterexamples of exhaustiveness).